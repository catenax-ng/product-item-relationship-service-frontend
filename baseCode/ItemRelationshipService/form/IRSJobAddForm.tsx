import { Box } from '@mui/material'
import { Button } from 'cx-portal-shared-components'
import { IRSRequestBody } from 'features/irs/types'
import { useState } from 'react'
import { useForm } from 'react-hook-form'
import { IRSJobAddFormSelectField } from './IRSJobAddFormSelectField'
import { IRSJobAddFormTextfield } from './IRSJobAddFormTextfield'
import { useDispatch } from 'react-redux'
import { useTranslation } from 'react-i18next'
import { fetchJobs, postJob } from 'features/irs/actions'

type DefaultFormFieldValuesType = {
  RequestBodyValues: string
  Environment: string
}

export const IRSJobAddForm = () => {
  const [response, setResponse] = useState<any>()
  const dispatch = useDispatch()
  const { t } = useTranslation()

  const testJob = {
    aspects: ['AssemblyPartRelationship', 'SerialPartTypization'],
    bomLifecycle: 'asBuilt',
    collectAspects: true,
    direction: 'downward',
    depth: 10,
    globalAssetId: 'urn:uuid:d387fa8e-603c-42bd-98c3-4d87fef8d2bb',
  }

  const defaultFormFieldValues = {
    RequestBodyValues: JSON.stringify(testJob, null, 2),
    Environment: 'https://irs.dev.demo.catena-x.net/irs/',
  }

  const {
    handleSubmit,
    getValues,
    control,
    trigger,
    formState: { errors },
  } = useForm({
    defaultValues: defaultFormFieldValues,
    mode: 'onChange',
  })

  const handleConfirm = async (formValues: DefaultFormFieldValuesType) => {
    console.log('Form data:', formValues)
    try {
      // TODO: Do Action Write to DB
      dispatch(postJob(JSON.parse(formValues.RequestBodyValues)))
      console.log()
      // setResponse(result)
      // dispatch(fetchJobs())
    } catch (error) {
      console.log(error)
    }
  }

  const onFormSubmit = async () => {
    const validateFields = await trigger(['RequestBodyValues', 'Environment'])

    if (validateFields) {
      const formValues = getValues()
      handleConfirm(formValues)
    }
  }

  const formHasErrors = () => {
    return Object.keys(errors).length > 0
  }

  return (
    <section>
      <Box className="irs-job-form">
        <Box className="irs-job-form-header">
          <h5>{t('content.irs.form.title')}</h5>
        </Box>
        <Box
          className="irs-job-form-content"
          style={{
            margin: '20px',
          }}
        >
          <form>
            <IRSJobAddFormTextfield
              {...{
                control,
                errors,
                trigger,
                name: 'RequestBodyValues',
                rules: {
                  required: true,
                  validate: (value: string) => {
                    try {
                      JSON.parse(value)
                    } catch (e) {
                      return false
                    }
                    return true
                  },
                },
                label: t('content.irs.form.requestBodyValues.label'),
                helperText: t('content.irs.form.requestBodyValues.helperText'),
              }}
            ></IRSJobAddFormTextfield>
            <IRSJobAddFormSelectField
              {...{
                control,
                errors,
                trigger,
                name: 'Environment',
                rules: {
                  required: true,
                  validate: (value: string) => value !== 'none',
                },
                label: t('content.irs.form.environment.label'),
              }}
            ></IRSJobAddFormSelectField>
          </form>
          <Button
            style={{ marginTop: 20 }}
            variant="contained"
            color="secondary"
            onClick={onFormSubmit}
            disabled={formHasErrors()}
          >
            {t('content.irs.form.sendButton')}
          </Button>
        </Box>
      </Box>
    </section>
  )
}
