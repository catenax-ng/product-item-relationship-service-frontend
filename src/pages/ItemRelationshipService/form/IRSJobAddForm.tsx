import { Box } from "@mui/material";

import { LoadingButton, PageSnackbar } from "cx-portal-shared-components";
import { useForm } from "react-hook-form";
import { useTranslation } from "react-i18next";
import { PaddedSection } from "../../../components/layout/PaddedSection";
import { getCurrentEnvironment, serverConfig } from "../../../constants/serverConfig";
import { useCreateJob } from "../../../services/queries/jobs";
import { IRSRequestBody } from "../../../types/jobs";
import { IRSJobAddFormTextfield } from "./components/IRSJobAddFormTextfield";

type DefaultFormFieldValuesType = {
  RequestBodyValues: string;
  Environment: string;
};

const useGetCurrentServerUrl = () => {
  const serverEnv = getCurrentEnvironment();
  return serverConfig[serverEnv].value;
};

export const IRSJobAddForm = () => {
  const { t } = useTranslation();
  const serverUrl = useGetCurrentServerUrl();
  const { mutate: createJob, isLoading, isError, isSuccess } = useCreateJob();

  const testJob = {
    aspects: ["AssemblyPartRelationship", "SerialPartTypization"],
    bomLifecycle: "asBuilt",
    collectAspects: true,
    direction: "downward",
    depth: 10,
    globalAssetId: "urn:uuid:d387fa8e-603c-42bd-98c3-4d87fef8d2bb",
  };

  const defaultFormFieldValues = {
    RequestBodyValues: JSON.stringify(testJob, null, 2),
    Environment: serverUrl,
  };

  const {
    handleSubmit,
    getValues,
    control,
    trigger,
    formState: { errors },
  } = useForm({
    defaultValues: defaultFormFieldValues,
    mode: "onChange",
  });

  const handleConfirm = async (formValues: DefaultFormFieldValuesType) => {
    try {
      const formData = JSON.parse(formValues.RequestBodyValues) as IRSRequestBody;
      createJob(formData);
    } catch (error) {
      console.log(error);
    }
  };

  const onFormSubmit = async () => {
    const validateFields = await trigger(["RequestBodyValues", "Environment"]);

    if (validateFields) {
      const formValues = getValues();
      handleConfirm(formValues);
    }
  };

  const formHasErrors = () => {
    return Object.keys(errors).length > 0;
  };

  return (
    <PaddedSection>
      <Box className="irs-job-form">
        <Box className="irs-job-form-header">
          <h5>{t("content.irs.form.title")}</h5>
        </Box>
        <Box
          className="irs-job-form-content"
          style={{
            margin: "20px",
          }}
        >
          <form onSubmit={handleSubmit(onFormSubmit)}>
            <IRSJobAddFormTextfield
              {...{
                control,
                errors,
                trigger,
                name: "RequestBodyValues",
                rules: {
                  required: true,
                  validate: (value: string) => {
                    try {
                      JSON.parse(value);
                    } catch (e) {
                      return false;
                    }
                    return true;
                  },
                },
                label: t("content.irs.form.requestBodyValues.label"),
                helperText: t("content.irs.form.requestBodyValues.helperText"),
              }}
            />
          </form>
          <LoadingButton
            style={{ marginTop: 20 }}
            color="secondary"
            label={t("content.irs.form.sendButton")}
            loadIndicator="Creating Job ..."
            loading={isLoading}
            onClick={onFormSubmit}
            disabled={formHasErrors()}
            onButtonClick={onFormSubmit}
            variant="contained"
            size="medium"
          />
        </Box>
      </Box>

      {isError && (
        <PageSnackbar
          description="Failed to create Job"
          open={true}
          severity="error"
          showIcon={true}
          title="Error"
          vertical={"top"}
          horizontal={"center"}
        />
      )}

      {isSuccess && (
        <PageSnackbar
          description="Created Job"
          open={true}
          severity="success"
          showIcon={true}
          title="Success"
          vertical={"top"}
          horizontal={"center"}
        />
      )}
    </PaddedSection>
  );
};
