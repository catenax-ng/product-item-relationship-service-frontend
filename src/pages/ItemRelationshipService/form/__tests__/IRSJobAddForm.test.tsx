import { render } from "../../../../testing/test-utils";
import { IRSJobAddForm } from "../IRSJobAddForm";

it("renders the form", () => {
  const { container } = render(<IRSJobAddForm />);
  expect(container).toMatchSnapshot();
});
