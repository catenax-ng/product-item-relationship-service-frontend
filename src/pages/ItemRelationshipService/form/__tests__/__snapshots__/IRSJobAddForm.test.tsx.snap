// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the form 1`] = `
<div>
  <section
    class="css-jlkg07"
  >
    <div
      class="irs-job-form MuiBox-root css-0"
    >
      <div
        class="irs-job-form-header MuiBox-root css-0"
      >
        <h5>
          content.irs.form.title
        </h5>
      </div>
      <div
        class="irs-job-form-content MuiBox-root css-0"
        style="margin: 20px;"
      >
        <form>
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-animated css-y7y8hp-MuiFormLabel-root-MuiInputLabel-root"
          >
            content.irs.form.requestBodyValues.label
          </label>
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <div
              class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-1ra8cd2-MuiInputBase-root-MuiFilledInput-root"
            >
              <textarea
                aria-describedby=":r0:-helper-text"
                aria-invalid="false"
                class="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputMultiline css-7209ej-MuiInputBase-input-MuiFilledInput-input"
                id=":r0:"
                maxlength="1000"
                style="height: 0px; overflow: hidden;"
              >
                {
  "aspects": [
    "AssemblyPartRelationship",
    "SerialPartTypization"
  ],
  "bomLifecycle": "asBuilt",
  "collectAspects": true,
  "direction": "downward",
  "depth": 10,
  "globalAssetId": "urn:uuid:d387fa8e-603c-42bd-98c3-4d87fef8d2bb"
}
              </textarea>
              <textarea
                aria-hidden="true"
                class="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputMultiline css-7209ej-MuiInputBase-input-MuiFilledInput-input"
                readonly=""
                style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
                tabindex="-1"
              />
            </div>
            <p
              class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained MuiFormHelperText-filled css-u5ud9e-MuiFormHelperText-root"
              id=":r0:-helper-text"
            >
              246/1000
            </p>
          </div>
        </form>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium css-zcbmsk-MuiButtonBase-root-MuiButton-root"
          style="margin-top: 20px;"
          tabindex="0"
          type="button"
        >
          content.irs.form.sendButton
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </section>
</div>
`;
